app/src/org/commcare/activities/QueryRequestActivity.java

```diff
@@ -0,0 +1,335 @@
+import org.commcare.tasks.templates.CommCareTaskConnector;
+import org.commcare.views.ManagedUi;
+import org.commcare.views.UiElement;
+import org.commcare.views.ViewUtil;
+import org.commcare.views.dialogs.CustomProgressDialog;
+import org.commcare.views.media.MediaLayout;
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.services.locale.Localization;
+import org.javarosa.core.services.locale.Localizer;
+import org.javarosa.xml.ElementParser;
+import org.javarosa.xml.TreeElementParser;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.Map;
+
+/**
+ * Collects 'query datum' in the current session. Prompts user for query
+ * params, makes query to server and stores xml 'fixture' response into current
+ * session. Allows for 'case search and claim' workflow when used inside a
+ * 'sync-request' entry in conjuction with entity select datum and sync
+ *
+ * @author Phillip Mates (pmates@dimagi.com).
+ */
+@ManagedUi(R.layout.http_request_layout)
+public class QueryRequestActivity
+        extends SaveSessionCommCareActivity<QueryRequestActivity>
+        implements HttpResponseProcessor {
+    private static final String TAG = QueryRequestActivity.class.getSimpleName();
+
+    private static final String ANSWERED_USER_PROMPTS_KEY = "answered_user_prompts";
+    private static final String IN_ERROR_STATE_KEY = "in-error-state-key";
+    private static final String ERROR_MESSAGE_KEY = "error-message-key";
+
+    @UiElement(value = R.id.request_button, locale = "query.button")
+    private Button queryButton;
+
+    @UiElement(value = R.id.error_message)
+    private TextView errorTextView;
+
+    private boolean inErrorState;
+    private String errorMessage;
+    private RemoteQuerySessionManager remoteQuerySessionManager;
+    private final Hashtable<String, EditText> promptsBoxes = new Hashtable<>();
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        remoteQuerySessionManager =
+                buildQuerySessionManager(CommCareApplication._().getCurrentSessionWrapper());
+
+        if (remoteQuerySessionManager == null) {
+            Log.e(TAG, "Tried to launch remote query activity at wrong time in session.");
+            setResult(RESULT_CANCELED);
+            finish();
+        } else {
+            loadStateFromSavedInstance(savedInstanceState);
+
+            setupUI();
+        }
+    }
+
+    private static RemoteQuerySessionManager buildQuerySessionManager(AndroidSessionWrapper sessionWrapper) {
```

Comment:
would argue that this would be better as a static method of `RemoteQuerySessionManager.java`